lightpath/path.py

```diff
@@ -27,6 +28,43 @@
 logger = logging.getLogger(__name__)
 
 
+class DeviceState(enum.Enum):
+    """Description of BeamStates"""
+    Removed = 0
+    Inserted = 1
+    Unknown = 2
+    Faulted = 3
+
+
+def find_device_state(device):
+    """
+    Report the state of a device
+
+    The device must implement ``.inserted`` and ``removed``.
+
+    Parameters
+    ----------
+    device : ophyd.Device
+
+    Returns
+    -------
+    state: DeviceState
+    """
+    # Gather device information
+    try:
+        _in, _out = device.inserted, device.removed
+    except Exception as exc:
+        logger.exception("Unable to determine device state for %r", device)
+        return DeviceState.Faulted
+    # Check state consistency and return proper Enum
+    if _in and not _out:
+        return DeviceState.Inserted
+    elif _out and not _in:
+        return DeviceState.Removed
+    else:
+        return DeviceState.Unknown
```

Comment:
This method is :+1: :+1:, great simple solution! Puts all of the error checking in one place and introduces a faulted state for the bad situations and broken code.

Does this mean we should have devices raise exceptions when they know they are in a faulted state?

Reply:
Hmmm. Can you give me an example of device we would want to do this? In my mind, `Faulted` would be reserved for a disconnection issue or something beyond just not being in any of the designated states. (In the `clear` method I opted to not even to try and call `.remove` on faulted devices)

Reply:
Ah, so a device that knows it's in a bad state either reports "in" and "out", or neither "in" nor "out"? Maybe I'm tripping over semantics for things like the `Attenuator` which has a status PV that shows faulted in some scenarios.

Reply:
Yeah maybe we should think about this, current logic is:
not in or not out -> "Unknown"
in and out -> "Unknown"
unable to report -> "Faulted"

That second one might want to be "Faulted" . My original thought was that we wanted to distinguish connection issues from just things being in the wrong place. Helpful for operators to be like this is "Unknown" I'm going to try and move this, and "Faulted" this IOC is down we need to call someone

Reply:
So our discussion was:
not connected -> disconnected
in and not out -> inserted
out and not in -> removed
not in and not out -> unknown
exception raised -> error
seem reasonable?